File: C_plus_plus/Binary_Tree_Maximum_Path_Sum.cpp
class Solution {
public:
    int max;
    int maxPathSum(TreeNode *root) {
        max = root->val;
        int temp = dp(root);
        return max>temp?max:temp;
    }
    int dp(TreeNode * node){
        if(node == NULL) return 0;
        int a = dp(node->left);
        int b = dp(node->right);
        int temp = a>b?a:b;
        if(a+b+node->val>max) max = a+b+node->val;
        if(node->val<0&&temp>max) max = temp;
        if(temp<=0) return node->val;
        else return node->val+temp;
    }
};

File: java/Binary_Tree_Maximum_Path_Sum.java
// DFS solution
// Calculate all the submax sum that contains node i, and compare with the global max.
// Time is O(n), space is O(n).

public class Solution {
    int max = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        findMax(root);
        return max;
    }
    public int findMax(TreeNode root) {
        if (root == null) return 0;
        int left = findMax(root.left);
        int right = findMax(root.right);
        int ret = Math.max(Math.max(left,right),0) + root.val;
        int pass = left + right + root.val;
        int submax = Math.max(ret,pass);
        if (submax > max) max = submax;
        return ret;
    }
}
